Adds specificity lesson

rachelandrew opened the pull request:
Fixes #4779 

Adding the Specificity lesson. There are a bunch of self-assessments in this which I'll add to the preflight issue as to do items.

una commented on src/site/content/en/learn/css/specificity/index.md:
```diff
@@ -0,0 +1,363 @@
+is the key to understanding how CSS decides between competing rules.
+
+Specificity is one of the four distinct stages of the cascade,
+which you learned about in the last module.
```
```suggestion
which was covered in the last module, on [the cascade](/learn/css/the-cascade/).
```

una commented on src/site/content/en/learn/css/specificity/index.md:
```diff
@@ -0,0 +1,363 @@
+the balancing act is making sure the CSS rules you expect to apply, actually *do apply,*
+while generally keeping scores low to prevent complexity.
+The score should only be as high as we need it to be,
+rather than going for the highest score possible.
```
```suggestion
rather than aiming for the highest score possible.
```

una commented on src/site/content/en/learn/css/specificity/index.md:
```diff
@@ -0,0 +1,363 @@
+
+An element (type) or pseudo-element selector gets **1 point of specificity** .
+
+### Type selector
```
```suggestion
#### Type selector
```

rachelandrew replied:
these are subheads of "scoring each selector type" so this change seems odd. Can you explain the reasoning?

una replied:
noticed that the organization (at least for the first bit) was:

### Type

Explanation of type, how many **points** type gives

#### Subtype

demo

#### Subtype

demo

rachelandrew replied:
I think I get what you mean now, I'm committing these so I can see them in the doc as I need to pull it down to add the aside about double classes.

una commented on src/site/content/en/learn/css/specificity/index.md:
````diff
@@ -0,0 +1,363 @@
+}
+```
+
+### Pseudo-element selector
````
```suggestion
#### Pseudo-element selector
```

una commented on src/site/content/en/learn/css/specificity/index.md:
```diff
@@ -0,0 +1,363 @@
+
+A class or pseudo-class selector gets **10 points of specificity**.
+
+Class selector:
```
```suggestion
#### Class selector
```

rachelandrew replied:
Then these ones could be an h4, as they are sub-sub heads because there is a h3 which is https://deploy-preview-5019--web-dev-staging.netlify.app/learn/css/specificity/#class-attribute-or-pseudo-class-selector

una commented on src/site/content/en/learn/css/specificity/index.md:
````diff
@@ -0,0 +1,363 @@
+}
+```
+
+Pseudo-class selector:
````
```suggestion
#### Pseudo-class selector
```

(Just updating these for more consistency with the above)

una commented on src/site/content/en/learn/css/specificity/index.md:
````diff
@@ -0,0 +1,363 @@
+}
+```
+
+Attribute selector:
````
```suggestion
#### Attribute selector
```

una commented on src/site/content/en/learn/css/specificity/index.md:
```diff
@@ -0,0 +1,363 @@
+In diagrams and specificity calculators,
+the specificity is often visualized like this:
+
+{% Img src="image/VbAJIREinuYvovrBzzvEyZOpw5w1/d1q5VKbFOQG2TBXoBtgm.png", alt="A diagram demonstrating most specific to least specific selectors", width="800", height="474" %}
```
TODO: ADD TO #5001

una commented on src/site/content/en/learn/css/specificity/index.md:
````diff
@@ -0,0 +1,363 @@
+```
+
+<figure class="w-figure">
+{% Codepen {
````
@robdodson @devnook would it be possible to have an option for which tabs to show? I.e. I think here, showing the CSS tab next to the result is useful. Current default just shows the default.

robdodson replied:
I believe that option exists ([handbook](https://web.dev/handbook/web-dev-components/#codepen)). To show CSS and results, you'd do the following:

```njk
{% Codepen {
  user: 'robdodson',
  id: 'GRroyyX',
  height: 300,
  theme: 'dark',
  tab: 'css,result',
  allow: ['geolocation']
} %}
```

rachelandrew replied:
Maybe this is a task we can do as a preflight, go through them all and make sure they each show the optimal arrangement of tabs?

una replied:
Sounds good, i'll make an issue to track this

una replied:
#5081

una commented on src/site/content/en/learn/css/specificity/index.md:
```diff
@@ -0,0 +1,363 @@
+because it has one class selector.
+
+If you want to give this rule a boost,
+repeat the class selector like this:
```
```suggestion
a little hack you could do is to repeat the class selector like this:
```

rachelandrew replied:
I'm not a fan of the term "hack" here, it's not a "hack", it's as per spec.

una replied:
The way it's written now in the content hierarchy makes it seem like it's common/the norm to write double classes like this. How can we call out that this isn't common behavior, though something to be aware of?

una reviewed (changes requested):
Just a few notes here! Overall LGTM

rachelandrew commented:
@una something odd is going on with headings here. The ones under https://deploy-preview-5019--web-dev-staging.netlify.app/learn/css/specificity/#class-attribute-or-pseudo-class-selector could be made h4, but I don't understand changing some of the others to h4 as it seems a bit random. The existing structure looks correct as shown by how it is displayed in the internal nav.

argyleink reviewed (approved):
this lesson reads the most like a podcast episode! overall its greatly enhanced by the examples. 

i can get lost in the headings, but i don't think that's this articles fault?

re: assessments, I'll add them in a PR after this has merged 👍🏻

rachelandrew commented:
I've added an aside re: double classes. I also added links to MDN at the first mention of each selector. I wanted to do it for :not then thought it made sense everywhere.
